src/script/interpreter.cpp

```diff
@@ -1613,68 +1614,64 @@ template <class T> class CTransactionSignatureSerializer {
- */
-template <class T> uint256 GetSequencesSHA256(const T &txTo) {
+/** Compute the double SHA256 of the concatenation of all nSequences of a tx. */
+uint256 GetBIP143SequencesHash(const std::vector<CTxIn> &vin) {
     CHashWriter ss(SER_GETHASH, 0);
-    for (const auto &txin : txTo.vin) {
+    for (const auto &txin : vin) {
         ss << txin.nSequence;
     }
-    return ss.GetSHA256();
+    return ss.GetHash();
 }
 
-/** Compute the (single) SHA256 of the concatenation of all txouts of a tx. */
-template <class T> uint256 GetOutputsSHA256(const T &txTo) {
+/** Compute the double SHA256 of the concatenation of all txouts of a tx. */
+uint256 GetBIP143OutputsHash(const std::vector<CTxOut> &vout) {
     CHashWriter ss(SER_GETHASH, 0);
-    for (const auto &txout : txTo.vout) {
+    for (const auto &txout : vout) {
         ss << txout;
     }
-    return ss.GetSHA256();
-}
-
-/** Compute the (single) SHA256 of the concatenation of all amounts spent by a
- * tx. */
-uint256 GetSpentAmountsSHA256(const std::vector<CTxOut> &outputs_spent) {
-    CHashWriter ss(SER_GETHASH, 0);
-    for (const auto &txout : outputs_spent) {
-        ss << txout.nValue;
-    }
-    return ss.GetSHA256();
-}
-
-/** Compute the (single) SHA256 of the concatenation of all scriptPubKeys spent
- * by a tx. */
-uint256 GetSpentScriptsSHA256(const std::vector<CTxOut> &outputs_spent) {
-    CHashWriter ss(SER_GETHASH, 0);
-    for (const auto &txout : outputs_spent) {
-        ss << txout.scriptPubKey;
-    }
-    return ss.GetSHA256();
+    return ss.GetHash();
 }
 
 } // namespace
 
 template <class T>
 PrecomputedTransactionData::PrecomputedTransactionData(
     const T &txTo, std::vector<CTxOut> &&spent_outputs) {
-    m_prevouts_single_hash = GetPrevoutsSHA256(txTo);
-    m_sequences_single_hash = GetSequencesSHA256(txTo);
-    m_outputs_single_hash = GetOutputsSHA256(txTo);
-    hashPrevouts = SHA256Uint256(m_prevouts_single_hash);
-    hashSequence = SHA256Uint256(m_sequences_single_hash);
-    hashOutputs = SHA256Uint256(m_outputs_single_hash);
+    hashPrevouts = GetBIP143PrevoutsHash(txTo.vin);
+    hashSequence = GetBIP143SequencesHash(txTo.vin);
+    hashOutputs = GetBIP143OutputsHash(txTo.vout);
     m_spent_outputs = std::move(spent_outputs);
-    m_spent_amounts_single_hash = GetSpentAmountsSHA256(m_spent_outputs);
-    m_spent_scripts_single_hash = GetSpentScriptsSHA256(m_spent_outputs);
+    if (m_spent_outputs.size() > 0) {
+        m_inputs_merkle_root =
+            TxInputsMerkleRoot(txTo.vin, m_inputs_merkle_height);
+        m_outputs_merkle_root =
+            TxOutputsMerkleRoot(txTo.vout, m_outputs_merkle_height);
+        m_amount_outputs_sum = Amount::zero();
+        for (const CTxOut &output : txTo.vout) {
+            m_amount_outputs_sum += output.nValue;
+        }
+        {
+            m_amount_inputs_sum = Amount::zero();
+            std::vector<uint256> spent_outputs_hashes;
+            spent_outputs_hashes.reserve(txTo.vin.size());
```

Comment:
Why shouldn't this be m_spent_outputs.size()? You're copying them. If there is an invariant that these two sizes need to be the same, then that should be asserted instead of implied.